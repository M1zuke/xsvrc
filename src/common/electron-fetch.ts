import { IpcRenderer } from '../index';

export type ElectronFetchArgs

export function electronFetch<T>(): Promise<T> {
  return IpcRenderer.invoke('fetch', (e,args: ElectronFetchArgs) => {

  })
}
