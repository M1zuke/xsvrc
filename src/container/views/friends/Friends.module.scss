@import '../../../common/_mixins';
@import '../../../common/_spacing';
@import '../../../common/_colors';

.Component {
  display: grid;
  grid-template-rows: repeat(2, autoSize()) auto;
  grid-gap: $innerSpacing;
  height: 100%;
  overflow: hidden;

  .FriendsCountWrapper {
    display: grid;
    grid-template-columns: auto autoSize();

    .ShortOverview {
      display: grid;
      grid-template-columns: repeat(4, autoSize());
      align-items: center;

      .AllOnlineFriendsCount {
        padding: $tinySpacing $smallSpacing;
        background-color: $colorBlue;
        color: black;
        font-weight: bold;
        border-top-left-radius: $standardBorderRadius;
        border-bottom-left-radius: $standardBorderRadius;
      }

      .FilteredFriendsCount {
        padding: $tinySpacing $smallSpacing;
        background-color: $color3;
        font-weight: bold;
        border-top-right-radius: $standardBorderRadius;
        border-bottom-right-radius: $standardBorderRadius;
        margin-right: $innerSpacing;
      }
    }

    .SpecialFilter {
      display: grid;
      grid-template-columns: repeat(3, autoSize());
      grid-gap: $innerSpacing;
      justify-content: flex-end;
    }
  }

  .FilterButtons {
    background-color: $color1;
    padding: $innerSpacing;
    border-radius: $standardBorderRadius;

    .SearchByName {
      margin-bottom: $innerSpacing;
    }

    .NormalFilter {
      display: grid;
      grid-template-columns: repeat(28, autoSize());
      grid-gap: $smallSpacing;
    }
  }

  .FriendsList {
    display: grid;
    grid-gap: $innerSpacing;
    grid-template-columns: repeat(10, 1fr);
  }
}

@media only screen and (max-width: 1610px) {
  .Component {
    .FriendsList {
      grid-template-columns: repeat(8, 1fr);
    }
  }
}
