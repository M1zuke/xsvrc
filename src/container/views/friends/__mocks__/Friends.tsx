import React from 'react';

export const Friends = jest.fn(() => <div data-testid="friends" />);
