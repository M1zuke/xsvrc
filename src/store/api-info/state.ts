import { Loadable } from '../reducer';

export type VRCApiInfo = {
  clientApiKey: string;
};

export type AppInfoState = Loadable<VRCApiInfo>;

export const INITIAL_API_INFO_STATE: AppInfoState = null;
