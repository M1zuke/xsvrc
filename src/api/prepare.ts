import { electronFetch, ElectronResult, ErrorType, RequestConfig } from '../common/electron-controls';
import { AppState } from '../store';
import { resetStoredCookies, setStoredCookies } from '../store/persisted/actions';
import { isErrorType, Loadable } from '../store/reducer';
import { AppDispatch } from '../thunk';

export type FetchResult<T> = Omit<ElectronResult<T>, 'cookies'>;

export function api(path: string): string {
  return `https://api.vrchat.cloud/api/1/${path}`;
}

export function isLoaded<T>(object: Loadable<T>): object is T {
  return !(
    object === null ||
    object === undefined ||
    object === 'loading' ||
    object === 'not-found' ||
    isErrorType(object)
  );
}

function ensureApiKey(config: RequestConfig, key: string): RequestConfig {
  if (key) {
    if (!config.params) {
      config.params = [];
    }
    config.params.push({ key: 'apiKey', value: key });
  }
  return config;
}

export async function prepare<T>(
  state: (() => AppState) | AppState,
  dispatch: AppDispatch,
  config: RequestConfig,
): Promise<FetchResult<T> | ErrorType> {
  try {
    const $state = typeof state === 'function' ? state() : state;
    const apiKey = $state.apiInfo;
    const storedCookies = $state.persisted.cookies;

    if (isLoaded(apiKey)) {
      config = ensureApiKey(config, apiKey.clientApiKey);
    }

    const result = await electronFetch<T>({ ...config, storedCookies: storedCookies });

    if (result.type === 'entity') {
      if (result.cookies.length !== 0) {
        dispatch(setStoredCookies(result.cookies));
      }
      return Promise.resolve({
        result: result.result,
        type: 'entity',
      });
    }
    if (result.type === 'error' && result.error.statusCode === 401) {
      dispatch(resetStoredCookies());
    }

    return Promise.resolve(result);
  } catch (error) {
    return Promise.resolve({ type: 'code-error', error: error as Error });
  }
}
