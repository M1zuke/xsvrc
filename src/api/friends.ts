import { resetStoredCookies } from '../store/cookies/actions';
import { setFriendInfo } from '../store/friends/action';
import { FriendInfo } from '../store/friends/state';
import { AppThunkAction } from '../thunk';
import { api, prepare } from './prepare';

const limit = 100;

export function friends(offline?: boolean, friendsInfo: FriendInfo[] = [], offset = 0): AppThunkAction<Promise<void>> {
  return async function (dispatch, getState) {
    const state = getState();

    if (state.friends === null) {
      dispatch(setFriendInfo('loading'));
    }

    const response = await prepare<FriendInfo[]>(state, dispatch, {
      url: api('auth/user/friends'),
      params: [
        { key: 'offset', value: offset },
        { key: 'n', value: limit },
        { key: 'offline', value: (!!offline).toString() },
      ],
    });

    if (response.type === 'entity') {
      const newFriendInfo = [...friendsInfo, ...response.result];
      if (response.result.length === limit) {
        return dispatch(friends(offline, newFriendInfo, offset + limit));
      }
      dispatch(setFriendInfo(newFriendInfo));
    } else {
      dispatch(resetStoredCookies());
    }
  };
}
