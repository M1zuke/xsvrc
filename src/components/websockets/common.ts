import { isLoaded } from '../../api/prepare';
import { UserInfo } from '../../api/types';
import { AppState } from '../../store';

export type Comparison<V> = {
  from: V;
  to: V;
};

export type UserComparison<K extends keyof UserInfo = keyof UserInfo> = Partial<{
  [P in K]: Comparison<UserInfo[P]>;
}>;

export function compareUsers(oldUser: UserInfo | null = null, newUser: UserInfo): UserComparison {
  if (oldUser === null) {
    return {};
  }
  return Object.assign(
    {},
    ...Object.keys(newUser).map((key) => {
      const k = key as keyof UserInfo;
      const v = oldUser[k];
      const nv = newUser[k];
      if (Array.isArray(v) && Array.isArray(nv)) {
        const arrayComparison = compareArray(v, nv);
        if (arrayComparison.added.length === 0 && arrayComparison.removed.length === 0) {
          return undefined;
        }
        return {
          [key]: {
            from: v,
            to: nv,
          },
        };
      }

      if (oldUser[k] !== undefined && oldUser[k] !== newUser[k]) {
        return {
          [key]: {
            from: oldUser[k],
            to: newUser[k],
          },
        };
      }
      return undefined;
    }),
  );
}

type ArrayComparison<T extends string> = {
  added: T[];
  removed: T[];
};

function compareArray<T extends string = string>(a: T[], b: T[]): ArrayComparison<T> {
  const differencesToA: T[] = [];
  b.forEach((v) => {
    if (!a.includes(v)) {
      differencesToA.push(v);
    }
  });

  const differencesToB: T[] = [];
  a.forEach((v) => {
    if (!b.includes(v)) {
      differencesToA.push(v);
    }
  });
  return { added: differencesToA, removed: differencesToB };
}

export function getFriendsAndOldUser(state: AppState, userId: string): [UserInfo[], UserInfo | null] {
  if (isLoaded(state.friends.friendInfo)) {
    return [Object.values(state.friends.friendInfo), state.friends.friendInfo[userId] || null];
  }
  return [[], null];
}
