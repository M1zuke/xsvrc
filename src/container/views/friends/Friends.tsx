import React, { ReactElement, useCallback, useEffect, useMemo, useState } from 'react';
import { useSelector } from 'react-redux';
import { isLoaded } from '../../../api/prepare';
import { useApi } from '../../../api/use-api';
import { Button } from '../../../components/button/Button';
import { Checkbox } from '../../../components/checkbox/Checkbox';
import { Content } from '../../../components/content/Content';
import { FriendOverview } from '../../../components/friend-overview/FriendOverview';
import { LoadableContent } from '../../../components/loadable-content/LoadableContent';
import { ScrollableContent } from '../../../components/scrollable-content/ScrollableContent';
import { messages } from '../../../i18n/en';
import { selectFriendInfo } from '../../../store/friends/selectors';
import styles from './Friends.module.scss';

export const CharacterFilters = [
  'ALL',
  '#',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
] as const;
export type CharacterFilter = typeof CharacterFilters[number];

export function Friends(): ReactElement {
  const { friends } = useApi();
  const { active, offline } = useSelector(selectFriendInfo);
  const [filter, setFilter] = useState<CharacterFilter>('ALL');
  const [showPrivate, setShowPrivate] = useState(true);
  const [showOffline, setShowOffline] = useState(false);

  const friendsInfos = useMemo(() => (showOffline ? offline : active), [active, offline, showOffline]);

  const friendInfoWithOrWithoutPrivate = useMemo(() => {
    if (isLoaded(friendsInfos) && !showPrivate) {
      return friendsInfos.filter((fi) => fi.location !== 'private');
    }
    return friendsInfos;
  }, [friendsInfos, showPrivate]);

  const disableFilterButton = useCallback(
    (key: CharacterFilter) => {
      if (!isLoaded(friendInfoWithOrWithoutPrivate)) {
        return true;
      }

      if (key === 'ALL') {
        return false;
      }

      if (key === '#') {
        const filteredFriendInfo = friendInfoWithOrWithoutPrivate.filter(
          (o) => !CharacterFilters.includes(o.displayName[0].toUpperCase() as CharacterFilter),
        );
        return filteredFriendInfo.length === 0;
      }

      const filteredFriendInfo = friendInfoWithOrWithoutPrivate.filter(
        (o) => o.displayName[0].toLowerCase() === key.toLowerCase(),
      );
      return filteredFriendInfo.length === 0;
    },
    [friendInfoWithOrWithoutPrivate],
  );

  const filteredFriendInfo = useMemo(() => {
    if (!isLoaded(friendInfoWithOrWithoutPrivate) || filter === 'ALL') {
      return friendInfoWithOrWithoutPrivate;
    }
    if (filter === '#') {
      return friendInfoWithOrWithoutPrivate.filter(
        (o) => !CharacterFilters.includes(o.displayName[0].toUpperCase() as CharacterFilter),
      );
    }
    return friendInfoWithOrWithoutPrivate.filter((fi) => fi.displayName[0].toUpperCase() === filter);
  }, [filter, friendInfoWithOrWithoutPrivate]);

  const filterButtons = useMemo(() => {
    return CharacterFilters.map((key) => {
      return (
        <Button
          key={`select-filter-button-${key}`}
          aria-label={`select-filter-button-${key}`}
          onClick={() => setFilter(key)}
          active={filter === key}
          disabled={disableFilterButton(key)}
        >
          {key}
        </Button>
      );
    });
  }, [disableFilterButton, filter]);

  useEffect(() => {
    friends(showOffline).finally();

    const interval = setInterval(() => {
      friends(showOffline).finally();
    }, 25000);

    return () => {
      clearInterval(interval);
    };
  }, [friends, showOffline]);

  const [filteredFriendsCount, friendsCount] = useMemo(() => {
    if (isLoaded(friendsInfos) && isLoaded(filteredFriendInfo)) {
      return [filteredFriendInfo.length, friendsInfos.length];
    }
    return [0, 0];
  }, [filteredFriendInfo, friendsInfos]);

  return (
    <div className={styles.Component}>
      <Content className={styles.FriendsCountWrapper}>
        <div className={styles.AllOnlineFriendsCount}>{friendsCount}</div>
        <div className={styles.FilteredFriendsCount}>{filteredFriendsCount}</div>
        {messages.Views.Friends.FriendsOnline}
      </Content>
      <Content className={styles.FilterButtons}>
        <div className={styles.NormalFilter}>{filterButtons}</div>
        <div className={styles.SpecialFilter}>
          <Checkbox label={messages.Views.Friends.ShowOffline} onClick={setShowOffline} value={showOffline} />
          <Checkbox label={messages.Views.Friends.ShowPrivate} onClick={setShowPrivate} value={showPrivate} />
        </div>
      </Content>
      <ScrollableContent>
        <LoadableContent data={filteredFriendInfo}>
          {(data) => (
            <div className={styles.FriendsList}>
              {data.map((friendInfo) => (
                <FriendOverview friendInfo={friendInfo} key={`Friends-Overview-${friendInfo.displayName}`} />
              ))}
            </div>
          )}
        </LoadableContent>
      </ScrollableContent>
    </div>
  );
}
