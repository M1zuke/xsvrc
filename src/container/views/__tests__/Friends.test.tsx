import { waitFor } from '@testing-library/dom';
import pretty from 'pretty';
import React from 'react';
import { api } from '../../../api/prepare';
import * as ElectronFetchModule from '../../../common/electron-fetch';
import { ElectronResult } from '../../../common/electron-fetch';
import { AuthenticatedUserInfo } from '../../../store/user/state';
import { render, withAuthentication } from '../../../test-utils';
import { Friends } from '../friends/Friends';
import USER_INFO_FETCH_RESULTS from '../../app/__tests__/user-info-fetch-results.json';

jest.mock('../../../common/electron-fetch');
const { electronFetch: FetchMock } = ElectronFetchModule as jest.Mocked<typeof ElectronFetchModule>;

describe('<Friends /> Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('fetches friends', async () => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<AuthenticatedUserInfo>>((resolve) => {
        resolve({
          type: 'entity',
          cookies: [],
          result: USER_INFO_FETCH_RESULTS as AuthenticatedUserInfo,
        });
      }),
    );

    const customResult = render(<Friends />, {
      state: withAuthentication({
        friends: 'loading',
      }),
    });

    await waitFor(() => {
      expect(FetchMock).toHaveBeenCalledWith({
        params: [
          { key: 'offline', value: 'false' },
          { key: 'apiKey', value: 'testApiKey' },
        ],
        storedCookies: [{ key: 'auth', url: 'testUrl', value: 'testAuthToken' }],
        url: api('auth/user/friends'),
      });
    });

    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
    expect(customResult.store.getState().friends).toEqual(USER_INFO_FETCH_RESULTS);
  });

  it('receives an error', async () => {
    FetchMock.mockReturnValueOnce(
      new Promise<ElectronResult<AuthenticatedUserInfo>>((resolve) => {
        resolve({
          // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
          // @ts-ignore
          type: 'error',
          message: '123',
        });
      }),
    );

    const customResult = render(<Friends />, { state: withAuthentication() });

    await waitFor(() => {
      expect(FetchMock).toHaveBeenCalled();
    });

    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
    expect(customResult.store.getState().cookies).toEqual([]);
  });
});
