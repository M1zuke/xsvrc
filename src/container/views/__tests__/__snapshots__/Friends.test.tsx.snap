// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Friends /> Component fetches friends 1`] = `"<div>Friends</div>"`;

exports[`<Friends /> Component receives an error 1`] = `"<div>Friends</div>"`;
